src/dag.rs

```diff
@@ -202,69 +146,101 @@ impl<G: GraphNode> Dag<G> {
 
     // Iterate all the descendants of the relevant node, visiting parents before
     // their children.
-    pub fn top_down_from(&self, id: &G::I) -> Option<TopDown<G>> {
-        Some(TopDown {
-            dag: self,
-            visit_stack: Vec::new(),
-            unvisited_roots: vec![*self.id_to_idx.get(id.borrow())?],
-        })
+    pub fn top_down_from(&self, id: &G::I) -> Option<impl Iterator<Item = &G>> {
+        // Mindlessly recycle `TopologicalSort`, just reverse it and BAM!
+        // The back-and-forth of iterators is super awkward but necessary,
+        // because we need something with `DoubleEndedIterator` trait (i.e. Vec::<_>).
+        // Maybe there's a better way.
+        Some(
+            TopologicalSort::new(&self, vec![*self.id_to_idx.get(id.borrow())?])
+                .into_iter()
+                .collect::<Vec<&G>>()
+                .into_iter()
+                .rev(),
+        )
     }
 }
 
+// Possible states of a node during DFS.
+#[derive(Clone, PartialEq, Eq)]
+enum NodeState {
+    New,    // Initial state of every node.
+    Opened, // Node visited (pushed to the stack) but not yet closed.
+    Closed, // Node popped from the stack, all descendants visited.
+}
+
+// Struct that iterates over the nodes reachable by a set of given sources
+// in topological order (https://en.wikipedia.org/wiki/Topological_sorting),
+// that is, every node comes _after_ all its children (yes, techinically
+// this is reverse toposort, but here it kinda makes sense to call it that
+// since edges are meant to represent dependencies).
+//
+// If cycles are present, `next()` will find one eventually and from then
+// on will return `None`.
```

Comment:
It's kinda theoretical since these aren't actually doc comments anyway, but without a bug it's impossible for users of this type to create one with cycles, so this should be documented on one of its private methods I think

Reply:
True, I moved to the comment on `next_or_error` and adapted.